torchgen/gen_functionalization_type.py

```diff
@@ -353,6 +339,132 @@ def emit_view_functionalization_body(
 """
 
 
+def maybe_create_output(f: NativeFunction, var_name: str) -> str:
+    if len(f.func.returns) == 0:
+        return ""
+    return_type = dispatcher.returns_type(f.func.returns).remove_const_ref().cpp_type()
+    return f"{return_type} {var_name} = "
+
+
+def wrap_propagate_mutations_and_return(
```

Comment:
I think this function is long enough that it deserves a high level description of the algorithm. (Underlying concern: is there really no way to simplify this, e.g., by sanding off more of the edge cases)

Reply:
One of the things that is making this function very annoying to review is that it is not compositional. I need to understand the details of the surrounding context to know exactly how I can interact with `inner_out_name`. The general concept with torchgen.api.translate is that if you pass in some ambient context into some code generation pipeline, that context comes with enough information so that you can just interact it directly (e.g., using translate), instead of having to reconstruct the type context by snooping around the NativeFunctions implementation.

If you look at torchgen/dest/register_dispatch_key.py or ufunc.py as a counterpoint, the codegen is also factored into a number of helpers, but each helper tries to define a single function in its entirety, and when there is complicated internal structure it tries its best to pass in a context and make use of translate.

Reply:
(Ended up splitting this function up into a bunch of helpers; hopefully it's less ugly now)